# Contents

1. Importing libraries and files

2. Data Wrangling
    - 2.1 Renaming Columns
    - 2.2 Creating a Data Dictionary
    - 2.3 Creating a subset of data

3. Exporting Dataframes

# 1. Importing Libraries and Files

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
path= r'C:\Users\spada\OneDrive\Data Analytics\02-2023 Instacart Basket Analysis'

In [3]:
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 2. Data Wrangling

In [9]:
df_ords['days_since_prior_order'].value_counts(dropna=False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [10]:
df_ords.shape

(3421083, 7)

## 2.1 Renaming Columns

In [6]:
df_ords.rename(columns={'order_dow': 'order_days_of_week'}, inplace=True)

In [7]:
df_ords.head()

order_id  user_id eval_set  order_number  order_days_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [8]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [9]:
df_ords['order_id'].dtype

dtype('O')

In [6]:
#Importing dataset departments.csv
df_dep=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [11]:
#Transposing df_dep
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [13]:
#Transposing df.dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [15]:
#new transposed df_dep dataframe
df_dep_t=df_dep.T

In [16]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
#data index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [19]:
#Take the first row of df_dep_T for the header
new_header = df_dep_t.iloc[0]

In [20]:
new_header

0    department
Name: department_id, dtype: object

In [21]:
# take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [22]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [23]:
# set the header row as the df header
df_dep_t_new.columns = new_header

In [24]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 2.2 Creating a Data Dictionary

In [25]:
#data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [27]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [28]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [29]:
#subset for df_prods for data only from the snack dept.
df_snacks =  df_prods[df_prods['department_id']==19]

In [30]:
#index code
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [31]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [32]:
#creating new dataframe of snacks
df_snacks =  df_prods[df_prods['department_id']==19]

In [33]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Task Procedures Exercise 4.4

In [36]:
#Observing the df_ords dataframe
df_ords

order_id  user_id eval_set  order_number  order_days_of_week  \
0        2539329        1    prior             1                   2   
1        2398795        1    prior             2                   3   
2         473747        1    prior             3                   3   
3        2254736        1    prior             4                   4   
4         431534        1    prior             5                   4   
...          ...      ...      ...           ...                 ...   
3421078  2266710   206209    prior            10                   5   
3421079  1854736   206209    prior            11                   4   
3421080   626363   206209    prior            12                   1   
3421081  2977660   206209    prior            13                   1   
3421082   272231   206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [37]:
# Change a numeric dataframe that does not need to be in analysis and change to a more suitable format
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [39]:
df_ords['user_id'].dtype

dtype('O')

In [41]:
# Renaming order_days_of_week variable
df_ords.rename(columns = {'order_days_of_week' : 'order_day_of_week'}, inplace = True)

In [42]:
df_ords.head()

order_id user_id eval_set  order_number  order_day_of_week  \
0  2539329       1    prior             1                  2   
1  2398795       1    prior             2                  3   
2   473747       1    prior             3                  3   
3  2254736       1    prior             4                  4   
4   431534       1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [43]:
# Busiest hour for placing orders
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

## Busiest hour of day for placing orders is 10am

In [44]:
#Data Dictionary
data_dict = df_dep_t_new.to_dict('index')

In [46]:
#Determine the meaning behind a value of 4 in the "department_id" column
print(data_dict.get('4'))

{'department': 'produce'}


In [47]:
#Breakfast Item Sales
df_breakfast =  df_prods[df_prods['department_id']==14]

In [48]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [51]:
# Find all observations that include items from the following departments: alcohol, deli, beverages, and meat/seafood
df_alcohol =  df_prods[df_prods['department_id']==5]

In [52]:
df_alcohol

product_id                               product_name  aisle_id  \
51             52                        Mirabelle Brut Rose       134   
118           119                     Chardonnay Paso Robles        62   
149           150                                  Brut Rosé       134   
233           234                          Tennessee Whiskey       124   
248           249             Pinot Grigio, California, 2010        62   
...           ...                                        ...       ...   
49548       49544                    Cabernet Sauvignon Wine        28   
49566       49562              Blanc De Noirs Sparkling Wine       134   
49610       49606                        Organic Natural Red        28   
49665       49661                                      Porto       134   
49688       49684  Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
51                 5    14.4  
118                5     5.5  
149                5    12.9  
233                5     3.1  
248                5     2.7  
...              ...     ...  
49548              5     9.8  
49566              5     5.7  
49610              5     6.8  
49665              5     8.2  
49688              5     5.3  

[1056 rows x 5 columns]

In [53]:
df_deli =  df_prods[df_prods['department_id']==20]

In [54]:
df_deli

product_id                                   product_name  aisle_id  \
48             49  Vegetarian Grain Meat Sausages Italian - 4 CT        14   
61             62        Premium Deli Oven Roasted Turkey Breast        96   
73             74                Artisan Chick'n & Apple Sausage        14   
84             85                           Soppressata Piccante        96   
108           109                         Grape Leaf Hummus Wrap        13   
...           ...                                            ...       ...   
49558       49554                          Roasted Garlic Hommus        67   
49564       49560    Selects Natural Slow Roasted Chicken Breast        96   
49585       49581                   Pinto Bean and Cheese Pupusa        13   
49609       49605                     Classic Hummus Family Size        67   
49642       49638           Organic Superfirm Vacuum Packed Tofu        14   

       department_id  prices  
48                20    10.1  
61                20    14.6  
73                20     3.0  
84                20     8.9  
108               20     7.1  
...              ...     ...  
49558             20    14.8  
49564             20    14.5  
49585             20    10.5  
49609             20     3.5  
49642             20     9.0  

[1322 rows x 5 columns]

In [55]:
df_beverages =  df_prods[df_prods['department_id']==7]

In [56]:
df_beverages

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
19             20  Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
...           ...                                             ...       ...   
49659       49655                                     Apple Cider        98   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
19                 7     6.0  
...              ...     ...  
49659              7    10.7  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  

[4365 rows x 5 columns]

In [58]:
df_meat_seafood =  df_prods[df_prods['department_id']==12]

In [59]:
df_meat_seafood

product_id                                       product_name  \
16             17                                  Rendered Duck Fat   
22             23                             Organic Turkey Burgers   
34             35     Italian Herb Porcini Mushrooms Chicken Sausage   
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
...           ...                                                ...   
49425       49421                                   Salame, Peppered   
49440       49436                              Imitation Crab Flakes   
49509       49505                                Hot Italian Sausage   
49655       49651                                       Beef Brisket   
49674       49670                                Beef Summer Sausage   

       aisle_id  department_id  prices  
16           35             12    17.1  
22           49             12     8.2  
34          106             12    15.1  
39          106             12    22.5  
83            7             12    24.3  
...         ...            ...     ...  
49425       106             12    12.2  
49440        15             12    23.5  
49509       106             12    15.8  
49655       122             12    20.7  
49674       106             12    19.2  

[907 rows x 5 columns]

## 2.3 Creating a subset of data

In [60]:
# Dinner Parties subset
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [61]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

## Total row count in last dataframe: 7650 rows x 5 columns

In [62]:
#Information on user_id 1
df_ords.head()

order_id user_id eval_set  order_number  order_day_of_week  \
0  2539329       1    prior             1                  2   
1  2398795       1    prior             2                  3   
2   473747       1    prior             3                  3   
3  2254736       1    prior             4                  4   
4   431534       1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [65]:
df_user1 =  df_ords[df_ords['user_id']=='1']

In [66]:
df_user1

order_id user_id eval_set  order_number  order_day_of_week  \
0   2539329       1    prior             1                  2   
1   2398795       1    prior             2                  3   
2    473747       1    prior             3                  3   
3   2254736       1    prior             4                  4   
4    431534       1    prior             5                  4   
5   3367565       1    prior             6                  2   
6    550135       1    prior             7                  1   
7   3108588       1    prior             8                  1   
8   2295261       1    prior             9                  1   
9   2550362       1    prior            10                  4   
10  1187899       1    train            11                  4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

In [67]:
# User1 stats
df_user1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# 3. Exporting Dataframes

In [68]:
# Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [69]:
#Export df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))